qt/widgets/common/src/MantidTreeModel.cpp

```diff
@@ -0,0 +1,265 @@
+#include "MantidAPI/AnalysisDataService.h"
+#include "MantidQtWidgets/Common/AlgorithmDialog.h"
+#include "MantidQtWidgets/Common/InterfaceManager.h"
+#include "MantidQtWidgets/Common/MantidTreeModel.h"
+#include "MantidQtWidgets/Common/MantidWSIndexDialog.h"
+#include <Poco/ActiveResult.h>
+#include <qcoreapplication.h>
+#include <QMessageBox>
+
+using namespace std;
+using namespace MantidQt;
+using namespace MantidWidgets;
+using namespace Mantid::API;
+
+namespace {
+Mantid::Kernel::Logger g_log("WorkspaceWidget");
+}
+
+MantidTreeModel::MantidTreeModel() {}
+
+// Data display and saving methods
+void MantidTreeModel::deleteWorkspaces(const QStringList &wsNames) {
+  try {
+    if (!wsNames.isEmpty()) {
+      auto alg = createAlgorithm("DeleteWorkspaces");
+      alg->setLogging(false);
+      std::vector<std::string> vecWsNames;
+      vecWsNames.reserve(wsNames.size());
+      foreach (auto wsName, wsNames) {
+        vecWsNames.push_back(wsName.toStdString());
+      }
+      alg->setProperty("WorkspaceList", vecWsNames);
+      executeAlgorithmAsync(alg);
+    }
+  } catch (...) {
+    QMessageBox::warning(nullptr, "", "Could not delete selected workspaces.");
+  }
+}
+
+void MantidTreeModel::renameWorkspace(QStringList wsName) {
+  // Determine the algorithm
+  QString algName("RenameWorkspace");
+  if (wsName.size() > 1)
+    algName = "RenameWorkspaces";
+
+  QHash<QString, QString> presets;
+  if (wsName.size() > 1) {
+    presets["InputWorkspaces"] = wsName.join(",");
+  } else {
+    presets["InputWorkspace"] = wsName[0];
+  }
+  showAlgorithmDialog(algName, presets);
+}
+
+// Algorithm Display and Execution Methods
+Mantid::API::IAlgorithm_sptr
+MantidTreeModel::createAlgorithm(const QString &algName, int version) {
+  Mantid::API::IAlgorithm_sptr alg;
+  try {
+    alg = Mantid::API::AlgorithmManager::Instance().create(
+        algName.toStdString(), version);
+  } catch (...) {
+    QString message = "Cannot create algorithm \"" + algName + "\"";
+    if (version != -1) {
+      message += " version " + QString::number(version);
+    }
+    QMessageBox::warning(nullptr, "MantidPlot", message);
```

Comment:
I've recently discovered that if you leave the title argument of `QMessageBox` as an empty string then it inherits the current main window title. This is quite useful for us as we then don't need to specify it here.